File: set7/7/main.py
import sys

def input():
    return sys.stdin.readline().rstrip()

N = int(input())
circle = []
for i in range(N):
    k, x, r = map(int, input().split())
    circle.append((r, x, k))
circle = sorted(circle)
circle.append((1000000000, 0, 0))

A, B = map(int, input().split())
idxA, idxB = 0, 0
while idxA <= N and circle[idxA][2] != A: idxA += 1
while idxB <= N and circle[idxB][2] != B: idxB += 1

L, R = [A], [B]
visited = [0] * 200005
visited[A] = True

ax, ar = circle[idxA][1], circle[idxA][0]
bx, br = circle[idxB][1], circle[idxB][0]
for i in range(idxA + 1, N + 1):
    cr, cx, ck = circle[i]
    if cr == ar: continue
    if abs(ax - cx) > abs(ar - cr): continue
    ax, ar = cx, cr
    if ck == B:
        break
    L.append(ck)
    visited[ck] = True
for i in range(idxB + 1, N + 1):
    cr, cx, ck = circle[i]
    if cr == br: continue
    if abs(bx - cx) > abs(br - cr): continue
    bx, br = cx, cr
    if visited[ck]:
        while L[-1] != ck: L.pop(-1)
        break
    R.append(ck)
    
print(len(L) + len(R))
print(" ".join(map(str, L)), end=' ')
print(" ".join(map(str, R[::-1])), end='')

File: set7/7/Main.java
import java.util.*;
import java.lang.*;
import java.io.*;

public class Main {
    public static void main(String[] args) {
        FastReader rd = new FastReader();

        int N = rd.nextInt();
        ArrayList<Tuple> circle = new ArrayList<>();
        for(int i = 0; i < N; ++i) {
            int k = rd.nextInt(), x = rd.nextInt(), r = rd.nextInt();
            circle.add(new Tuple(r, x, k));
        }
        Collections.sort(circle);
        circle.add(new Tuple(1000000000, 0, 0));
        int A = rd.nextInt(), B = rd.nextInt();
        int idxA = 0, idxB = 0;
        while(idxA <= N && circle.get(idxA).c != A) ++idxA;
        while(idxB <= N && circle.get(idxB).c != B) ++idxB;

        Stack<Integer> L = new Stack<>();
        Stack<Integer> R = new Stack<>();
        boolean []visited = new boolean[200005];

        L.push(A);
        visited[A] = true;
        int ax = circle.get(idxA).b, ar = circle.get(idxA).a;
        int bx = circle.get(idxB).b, br = circle.get(idxB).a;
        for(int i = idxA + 1; i <= N; ++i) {
            Tuple value = circle.get(i);
            if(value.a == ar) continue;
            if(Math.abs(ax - value.b) > Math.abs(ar - value.a)) continue;
            ax = value.b; ar = value.a;
            if(value.c == B) break;
            L.push(value.c);
            visited[value.c] = true;
        }
        R.push(B);
        for(int i = idxB + 1; i <= N; i++) {
            Tuple value = circle.get(i);
            if(value.a == br) continue;
            if(Math.abs(bx - value.b) > Math.abs(br - value.a)) continue; 
            bx = value.b; br = value.a;
            if(visited[value.c]) {
                while(L.peek() != value.c) L.pop();
                break;
            }
            R.push(value.c);
        }
        System.out.println(L.size() + R.size());
        for(Integer x : L) System.out.print(x + " ");
        Collections.reverse(R);
        for(Integer x : R) System.out.print(x + " ");
    }

    static public class Tuple implements Comparable<Tuple> {
        int a, b, c;
        public Tuple(int a, int b, int c) {
            this.a = a;
            this.b = b;
            this.c = c;
        }
        public int compareTo(Tuple o) {
            if(o.a != this.a) return Integer.compare(this.a, o.a);
            return Integer.compare(this.b, o.b);
        }
    }

    static class FastReader {
        BufferedReader br;
        StringTokenizer st;

        public FastReader() {
            br = new BufferedReader(new InputStreamReader(System.in));
        }

        String next() {
            while(st == null || !st.hasMoreElements()) {
                try {
                    st = new StringTokenizer(br.readLine());
                }
                catch (IOException e) {
                    e.printStackTrace();
                }
            }
            return st.nextToken();
        }

        int nextInt() { return Integer.parseInt(next()); }
        long nextLong() { return Long.parseLong(next()); }
        double nextDouble() { return Double.parseDouble(next()); }
        String nextLine() {
            String str = "";
            try {
                str = br.readLine();
            }
            catch (IOException e) {
                e.printStackTrace();
            }
            return str;
        }
    }
}